paymentsheet/src/androidTest/java/com/stripe/android/paymentsheet/GooglePayTest.kt

```diff
@@ -38,52 +60,189 @@ internal class GooglePayTest {
         hasGooglePayConfig = true,
-    ) {
-        composeTestRule.onGooglePayOption().assertExists()
-    }
+        shouldHaveGooglePayOption = true,
+    )
 
     @Test
-    fun googlePayIsNotShownWhenNotReady() = runGooglePayTest(
+    fun googlePayIsNotShownWhenNotReady() = runGooglePayAvailabilityTest(
         isGooglePayReady = false,
         isGooglePayEnabledInElementsSession = true,
         hasGooglePayConfig = true,
-    ) {
-        composeTestRule.onGooglePayOption().assertDoesNotExist()
-    }
+        shouldHaveGooglePayOption = false,
+    )
 
     @Test
-    fun googlePayIsNotShownWhenNotEnabledInElementsSession() = runGooglePayTest(
+    fun googlePayIsNotShownWhenNotEnabledInElementsSession() = runGooglePayAvailabilityTest(
         isGooglePayReady = true,
         isGooglePayEnabledInElementsSession = false,
         hasGooglePayConfig = true,
-    ) {
-        composeTestRule.onGooglePayOption().assertDoesNotExist()
-    }
+        shouldHaveGooglePayOption = false,
+    )
 
     @Test
-    fun googlePayIsNotShownWhenGooglePayConfigIsNotProvided() = runGooglePayTest(
+    fun googlePayIsNotShownWhenGooglePayConfigIsNotProvided() = runGooglePayAvailabilityTest(
         isGooglePayReady = true,
         isGooglePayEnabledInElementsSession = true,
         hasGooglePayConfig = false,
+        shouldHaveGooglePayOption = false,
+    )
+
+    @Test
+    fun googlePaySucceeds() {
+        var resultCallbackCalled = false
+
+        runGooglePayFlowTest(
+            paymentResultCallback = {
+                resultCallbackCalled = true
+
+                assertCompleted(it)
+            }
+        ) { scenario ->
+            val paymentMethod = PaymentMethodFactory.card()
+
+            intendingGooglePayToBeLaunched(GooglePayPaymentMethodLauncher.Result.Completed(paymentMethod))
+            intendingPaymentConfirmationToBeLaunched(
+                InternalPaymentResult.Completed(PaymentIntentFactory.create(paymentMethod))
+            )
+
+            scenario.confirm()
+
+            intendedGooglePayToBeLaunched()
+            intendedPaymentConfirmationToBeLaunched()
+
+            composeTestRule.waitUntil(UI_TIMEOUT) {
+                resultCallbackCalled
+            }
+        }
+    }
+
+    @Test
+    fun googlePayFailsOnGooglePaySheetFailure() = runGooglePayFlowTest(
+        paymentResultCallback = {
+            if (integrationType == ProductIntegrationType.FlowController) {
+                assertFailed(it)
+            }
+        }
+    ) { scenario ->
+        intendingGooglePayToBeLaunched(
+            GooglePayPaymentMethodLauncher.Result.Failed(
+                error = IllegalStateException("An error occurred!"),
+                errorCode = GooglePayPaymentMethodLauncher.INTERNAL_ERROR
+            )
+        )
+
+        scenario.confirm()
+
+        intendedGooglePayToBeLaunched()
+
+        if (integrationType == ProductIntegrationType.PaymentSheet) {
```

Comment:
I don't love the forking on different integration shapes. It might make me feel better if we asserted something different for flow controller.

Reply:
Should we write different tests for `PaymentSheet` & `FlowController`?